File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/com/example/factory/factories/DocumentFactory.java
package com.example.factory.factories;
import com.example.factory.documents.Document;
public abstract class DocumentFactory{
    public abstract Document createDocument();
    public Document processDocument()
    {
        System.out.println("=== Document Processing Started ===");
        Document document = createDocument();
        System.out.println("Document Type: " + document.getDocumentType());
        System.out.println("Creation Time: " + document.getCreationTime());
        document.open();
        document.save();
        System.out.println("=== Document Processing Completed ===");
        return document;
    }    
}

File: Week1_Design Patterns and Principles/Ex_2_Implementing_the_Factory_Method_Pattern/Code/Factories/DocumentFactory.cs
using System;
using FactoryDemo.Documents;

namespace FactoryDemo.Factories;

public abstract class DocumentFactory
{
    public abstract IDocument CreateDocument();


    public IDocument ProcessDocument()
    {
        Console.WriteLine("=== Document Processing Started ===");
        var doc = CreateDocument();
        Console.WriteLine($"Document Type: {doc.GetDocumentType()}");
        Console.WriteLine($"Creation Time: {doc.GetCreationTime()}");
        doc.Open();
        doc.Save();
        Console.WriteLine("=== Document Processing Completed ===");
        return doc;
    }
}
